File: back-end/src/main/java/Individual.java
package main.java;

import java.util.ArrayList;

public class Individual extends User {

    private int balance;
    private String displayName;
    private boolean isAdmin;

    private ArrayList<StringStringPair> paidHouses = new ArrayList<StringStringPair>();

    public Individual(String username, int balance, String displayName, boolean isAdmin) {
        super(username);
        if (balance < 0)
            balance = 0;
        this.balance = balance;
        this.displayName = displayName;
        this.isAdmin = isAdmin;
    }

    public Individual(String username, int balance, String displayName, boolean isAdmin,
                      ArrayList<StringStringPair> paidHouses) {
        this(username, balance, displayName, isAdmin);
        this.paidHouses = new ArrayList<StringStringPair>(paidHouses);
    }

    public boolean getIsAdmin() {
        return this.isAdmin;
    }

    public boolean hasPaidforHouse(String houseId, String ownerId) {
        for (StringStringPair member : paidHouses) {
            if (member.equals(new StringStringPair(ownerId, houseId))) {
                return true;
            }
        }
        return false;
    }

    private static int phonePrice = 1000;

    public boolean payForHouse(String houseId, String ownerId) {
        if (balance >= phonePrice) {
            paidHouses.add(new StringStringPair(ownerId, houseId));
            balance -= phonePrice;
            return true;
        }
        return false;
    }

    public int getBalance() {
        return balance;
    }

    public void setBalance(int balance) {
        if (balance < 0)
            balance = 0;
        this.balance = balance;
    }

    public String getDisplayName() {
        return displayName;
    }

    ArrayList<StringStringPair> getPaidHouses() {
        return paidHouses;
    }
}

File: back-end-prototype/models/individual.js
'use strict';
const models = require('./index');
const PHONE_PRICE = 1000;

module.exports = (sequelize, DataTypes) => {
    var Individual = sequelize.define(
        'Individual',
        {
            username: {
                type: DataTypes.STRING,
                get() {
                    return this.getDataValue('username');
                }
            },
            displayName: {
                type: DataTypes.STRING,
                get() {
                    return this.getDataValue('displayName');
                }
            },
            balance: {
                type: DataTypes.INTEGER,
                set(newBalance) {
                    if (newBalance < 0) newBalance = 0;
                    this.setDataValue('balance', newBalance);
                }
            },
            isAdmin: {
                type: DataTypes.BOOLEAN,
                get() {
                    return this.getDataValue('isAdmin');
                }
            }
        },
        {
            getterMethods: {
                async hasPaidForHouse(houseId, ownerId) {
                    let paidHouses = await models.PaidHouse.findAll({
                        where: {
                            individualId: this.getDataValue('username')
                        }
                    });
                    let i;
                    for (i = 0; i < paidHouses.length; i++) {
                        if (
                            paidHouses[i].getDataValue(houseId) === houseId &&
                            paidHouses[i].getDataValue(ownerId) === ownerId
                        )
                            return true;
                    }
                    return false;
                },
                json() {
                    let res = {};
                    res.username = this.username;
                    res.balance = this.balance;
                    res.name = this.displayName;
                    return res;
                }
            },
            setterMethods: {
                async payForHouse(houseId, ownerId) {
                    try {
                        this.balance -= PHONE_PRICE;
                        let paidBefore = await this.hasPaidForHouse(
                            houseId,
                            ownerId
                        );
                        if (!paidBefore)
                            await models.PaidHouse.create({
                                individualId: this.username,
                                ownerId: ownerId,
                                houseId: houseId
                            });
                        return true;
                    } catch (err) {
                        return false;
                    }
                }
            }
        }
    );
    Individual.associate = function(models) {
        // associations can be defined here
    };
    return Individual;
};
